_posts/2019-05-24-use-react-hooks-with-storage-as-global-state-management.md

```diff
@@ -0,0 +1,163 @@
+
+```js
+import RTStorage from 'rt-storage';
+
+const storage = new RTStorage({ name: 'test-db' });
+const subscription = storage.subscribe((event) => {
+  console.dir(event)
+});
+const storageKey = 'storageKey';
+storage.getItem(storageKey).then(a => console.log(a));
+storage.setItem(storageKey, data).then(() => console.log('set successfully'));
+subscription.unsubscribe();
+```
+
+For more details, please check the project [repo](https://github.com/embbnux/rt-storage).
+
+## Use storage hook
+
+So let’s start to create a customized react hook with global storage. Following is the core source code of [use-global-storage](https://github.com/embbnux/use-global-storage):
+
+```js
+import { useState, useEffect } from 'react';
+import * as RTStorage from 'rt-storage';
+export default function useGlobalStorage({ storageOptions } : { storageOptions: any }) {
+  const storage = new RTStorage(storageOptions);
+  const useStorage = (key: string, initialData: any) => {
+    const [data, setState] = useState(initialData);
+    useEffect(() => {
+      function handleStorageChange(data) {
+        setState(data);
+      }
+      storage.getItem(key).then(lastData => {
+        if (lastData) {
+          setState(lastData);
+        }
+      });
+      const subscription = storage.subscribe(key, handleStorageChange);
+      return () => {
+        subscription.unsubscribe();
+      };
+    }, []);
+    const setData = async(newData: any) => {
+      let newValue;
+      if (typeof newData === 'function') {
+        newValue = newData(data);
+      } else {
+        newValue = newData
+      }
+      setState(newValue);
+      await storage.setItem(key, newValue);
+    }
+  
+    return [data, setData];
+  }
+  return useStorage;
+};
+```
+
+It will return a hook function which allow user to set state into storage and sync for other components and tabs. setData function will set state into react state and save data into storage. And we use `useEffect` to subscribe storage changed event.
+
+Example to use `use-global-storage` hook in project:
+
+```js
+import useGlobalStorage from 'use-global-storage';
+
+const useStorage = useGlobalStorage({
+  storageOptions: { name: 'test-db' }
+});
+
+const Counter = () => {
+  const [state, setState] = useStorage('counter');
+  return (
+    <div className="Counter">
+      <p>
+        Counter:
+        {state || 0}
+      </p>
+      <button type="button" onClick={() => setState(state + 1)}>
+        +1 to global
+      </button>
+    </div>
+  );
+};
+```
+
+It is very simple to use it. Just get `state` and `setState` from `useStorage` hook, and your component is connected with storage data. You can get the online demo page [here](https://embbnux.github.io/use-global-storage-demo/). To get full code in this Github [repo](https://github.com/embbnux/use-global-storage-demo).
```

Comment:
It is very simple to use `hooks`. (Avoid using `it` like this. Unless you have already referred to something in the last sentence, so that readers can easily tell what `it` is referring to)
To get full code in this Github repo => To get the full code, refer to this Github repo